src/Microsoft.OData.Core/Json/BufferingJsonReader.cs

```diff
@@ -235,6 +249,140 @@ public bool Read()
             return this.ReadInternal();
         }
 
+        /// <summary>
+        /// The value of the last reported node.
+        /// </summary>
+        /// <remarks>
+        /// Depending on whether buffering is on or off this will return the node value of the last
+        /// buffered read or the node value of the last unbuffered read.
+        /// </remarks>
+        /// <returns>A task that represents the asynchronous read operation.
+        /// The value of the TResult parameter contains the node value of the last buffered read if buffering is on or off;
+        /// otherwise the node value of the last unbuffered read.</returns>
+        public Task<object> GetValueAsync()
+        {
+            this.AssertAsynchronous();
+
+            if (this.bufferedNodesHead != null)
+            {
+                if (this.isBuffering)
+                {
+                    Debug.Assert(this.currentBufferedNode != null, "this.currentBufferedNode != null");
+                    return Task.FromResult(this.currentBufferedNode.Value);
+                }
+
+                // in non-buffering mode if we have buffered nodes satisfy the request from the first node there
+                return Task.FromResult(this.bufferedNodesHead.Value);
+            }
+
+            return this.asyncInnerReader.GetValueAsync();
+        }
+
+        /// <summary>
+        /// Asynchronously reads the next node from the input.
+        /// </summary>
+        /// <returns>A task that represents the asynchronous read operation.
+        /// The value of the TResult parameter contains true if a new node was found, or false if end of input was reached.</returns>
+        public Task<bool> ReadAsync()
+        {
+            Debug.Assert(!this.parsingInStreamError, "!this.parsingInStreamError");
+            this.AssertAsynchronous();
+
+            // read the next node and check whether it is an in-stream error
+            return this.ReadInternalAsync();
+        }
+
+        /// <summary>
+        /// Asynchronously checks whether or not the current value can be streamed.
+        /// </summary>
+        /// <returns>A task that represents the asynchronous operation.
+        /// true if the current value can be streamed; otherwise false.</returns>
+        public virtual async Task<bool> CanStreamAsync()
+        {
+            this.AssertAsynchronous();
+
+            if (!this.isBuffering && this.asyncInnerReader is IJsonStreamReaderAsync asyncStreamReader)
+            {
+                return await asyncStreamReader.CanStreamAsync()
+                    .ConfigureAwait(false);
+            }
+
+            object value = await this.GetValueAsync()
+                .ConfigureAwait(false);
+            return value is string || value == null || this.NodeType == JsonNodeType.StartArray || this.NodeType == JsonNodeType.StartObject;

```

Comment:
I don't think the subsequent conditions are predicated on this first one for any reason, so I'm wondering if it makes sense to do the cast last due to short circuit evaluation

Reply:
@corranrogue9 Does a cast actually happen here given that `value` is already boxed and `string` is a reference type? In addition, supposing `StartArray` and `StartObject` occur rarely, do we not end up always doing the "cheap" comparisons as well as evaluating `value is string` expression?